### DISCUSSION TITLE
Whether to use router = APIRouter() uniformly?

### DISCUSSION THREAD

**Discussion by wu-clan** [fastapi-practices/fastapi_best_architecture#4] [Ideas]
Hi, @downdawn 

Reopen this question.

It is known to be unfriendly to IDE smart prompts.

Ensure that the \__init__ file only contains package import, and do not use it to implement code unless necessary.

**Comment by wu-clan** [comment]
We can use `as` to distinguish them.

E.g.:

```python
from backend.app.api.v1.user import router as user_router
```

**Comment by downdawn** [comment]
Suppose the ideal directory structure is like this:

api
├── auth
│ ├── __init__.py
│ ├── user.py
│ ├── service.py
├──__init__.py

api/__init__.py is equivalent to routers.py, here only need to include the router of all subdirectories

```
from fastapi import APIRouter

from api.auth import router as auth_router

v1 = APIRouter(prefix='/v1')

v1.include_router(auth_router)
...
```

api/auth/__init__.py

```
from fastapi import APIRouter
from api.auth.user import router as user_router

router = APIRouter(prefix='/auth', tags=['auth'])

router.include_router(user_router, prefix='/user')
...
```

But it is not well compatible with cvb, the prefix must be filled in user.py.

**Comment by wu-clan** [comment]
I have submitted the dev branch, please check it out and work on it first

**Reply by downdawn** [reply]
> I have submitted the dev branch, please check it out and work on it first我已经提交了dev分支，请先查看并处理

I run the dev branch locally, optimizing some issues.

I don't know whether to directly mention PR, or let's discuss it first.

1. Package guide problem.

I removed all backend.app, then python main.py can run normally.

2. Environment

I recommend using python-dotenv to manage the environment, which is also supported by pydantic.

At the same time, the official line is distinguished and the document is not displayed to ensure security.

```
    ENVIRONMENT: str
    # FastAPI
    TITLE: str = 'FastAPI'
    VERSION: str = 'v1'
    DESCRIPTION: str = "FastAPI Best Architecture"
    DOCS_URL: str = None
    REDOCS_URL: str = None
    OPENAPI_URL: str = None

    @root_validator
    def validator_api_url(cls, values):
        # Documents are only displayed in the local environment and test environment
        if values["ENVIRONMENT"] in ("local", "testing"):
            values["DOCS_URL"] = f'/{values["VERSION"]}/docs'
            values["REDOCS_URL"] = f'/{values["VERSION"]}/redocs'
            values["OPENAPI_URL"] = f'/{values["VERSION"]}/openapi'
        return values
   ...
    class Config:
        env_file = '.env'

Provide .env and env.example files
```

3. Redundant files

There are some files in utils that are not used.

Can `#!/usr/bin/env python3` be omitted.

4. docker

Normal back-end business will not configure the database and nginx in docker, right?

Whether to keep the most basic backend environment here, it is recommended to use [uvicorn-gunicorn-fastapi](https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker)

**Reply by wu-clan** [reply]
I think it can be directly PR. Comment directly on the code.

**Reply by downdawn** [reply]
> I think it can be directly PR. Comment directly on the code.

I create a new pr, please check it out.

There are a few more issues that I didn't fix directly:

1. Use the Dockerfile of uvicorn-gunicorn-fastapi

```
FROM tiangolo/uvicorn-gunicorn-fastapi:python3.10

COPY requirements.txt /app/requirements.txt

RUN pip install --no-cache-dir --upgrade -r /app/requirements.txt

COPY . /app

COPY app/.env.pro /app/.env

ENV ACCESS_LOG /app/logs/access.log
ENV ERROR_LOG /app/logs/error.log
```

2. Whether to delete ` #!/usr/bin/env python3` globally

3. Some interface exceptions

```
    @staticmethod
    async def get_user_info(username: str):
        async with async_db_session() as db:
            user = await UserDao.get_user_by_username(db, username)
            if not user:
                raise errors.NotFoundError(msg='用户不存在')
        if user.avatar is not None:
            user.avatar = cut_path(AvatarPath + user.avatar)[1]
        # here
        return user
```
4. Whether to simplify user-related interfaces and remove codes not used by utils
5. Add necessary tests, such as creating conftest and testing user-related interfaces

**Reply by wu-clan** [reply]
I will create several new issues 
1. Cannot change the base file, docker_compose.yal relies on Dockerfile
3. do not need to delete 
3. fix 
5. agree 
5. agree

**Comment by downdawn** [comment]
Hi, @wu-clan 
Add company.py under the auth directory, how to set the route? The current main branch does not seem to support

**Reply by wu-clan** [reply]
Don't add a router to `company.py`, it should be import from `user.py`

**Reply by wu-clan** [reply]
Only one router is allowed in one directory，because, since you put them together, they should belong to the same tag, otherwise create a new directory.

**Reply by downdawn** [reply]
If this is the case, the auth directory is not very useful, just be the same as task_demo

**Reply by wu-clan** [reply]
For projects that are not large, it is good to have all routing files in one level of the directory.

> Don't add a router to `company.py`, it should be import from `user.py`

This does lead to certain problems if using the class view.

Do you have a program yet?

**Reply by downdawn** [reply]
Yes, this is not very scalable

A pseudocode has been written above

```
from fastapi import APIRouter
from api.auth.user import router as user_router

router = APIRouter(prefix='/auth', tags=['auth'])

router.include_router(user_router, prefix='/user')
...
```

**Reply by wu-clan** [reply]
I'm working for #23 before I start router

**Comment by wu-clan** [comment]
Hi, @downdawn 

Please check：#32
